library/common/http/dispatcher.cc

```diff
@@ -93,134 +102,93 @@ void Dispatcher::DirectStreamCallbacks::encodeHeaders(const ResponseHeaderMap& h
-    // **not** have a dangling reference.
-    ENVOY_LOG(debug, "[S{}] scheduling cleanup", direct_stream_.stream_handle_);
-    http_dispatcher_.cleanup(direct_stream_.stream_handle_);
-  }
+void Dispatcher::DirectStreamCallbacks::closeStream() {
+  // Envoy itself does not currently allow half-open streams where the local half is open
+  // but the remote half is closed. Note that if local is open, Envoy will reset the stream.
+  http_dispatcher_.removeStream(direct_stream_.stream_handle_);
 }
 
-Stream& Dispatcher::DirectStreamCallbacks::getStream() { return direct_stream_; }
+void Dispatcher::DirectStreamCallbacks::onComplete() {
+  ENVOY_LOG(debug, "[S{}] complete stream (success={})", direct_stream_.stream_handle_, success_);
+  if (success_) {
+    http_dispatcher_.stats().stream_success_.inc();
+  } else {
+    http_dispatcher_.stats().stream_failure_.inc();
+  }
+  bridge_callbacks_.on_complete(bridge_callbacks_.context);
+}
 
 void Dispatcher::DirectStreamCallbacks::onReset() {
   ENVOY_LOG(debug, "[S{}] remote reset stream", direct_stream_.stream_handle_);
+
+  // The stream should no longer be preset in the map, because onReset() was either called from a
+  // terminal callback that mapped to an error or it was called in response to a resetStream().
+  ASSERT(!http_dispatcher_.getStream(direct_stream_.stream_handle_));
   envoy_error_code_t code = error_code_.value_or(ENVOY_STREAM_RESET);
   envoy_data message = error_message_.value_or(envoy_nodata);
   int32_t attempt_count = error_attempt_count_.value_or(-1);
 
   // Testing hook.
   http_dispatcher_.synchronizer_.syncPoint("dispatch_on_error");
 
-  // direct_stream_ will not be a dangling reference even in the case that closeRemote cleaned up
-  // because in that case this reset is happening synchronously, with the encoding call that called
-  // closeRemote, in the Envoy Main thread. Hence DirectStream destruction which is posted on the
-  // Envoy Main thread's event loop will strictly happen after this direct_stream_ reference is
-  // used. @see Dispatcher::DirectStreamCallbacks::closeRemote() for more details.
-  if (direct_stream_.dispatchable(true)) {
-    ENVOY_LOG(debug, "[S{}] dispatching to platform remote reset stream",
-              direct_stream_.stream_handle_);
-    // Only count the on error if it is dispatchable. Otherwise, the onReset was caused due to a
-    // client side cancel via Dispatcher::DirectStream::resetStream().
-    http_dispatcher_.stats().stream_failure_.inc();
-    bridge_callbacks_.on_error({code, message, attempt_count}, bridge_callbacks_.context);
-
-    // All the terminal callbacks only cleanup if they are dispatchable.
-    // This ensures that cleanup will happen exactly one time.
-    http_dispatcher_.cleanup(direct_stream_.stream_handle_);
-  }
+  ENVOY_LOG(debug, "[S{}] dispatching to platform remote reset stream",
+            direct_stream_.stream_handle_);
+  http_dispatcher_.stats().stream_failure_.inc();
```

Comment:
I wonder if we should move stats charging to closeStream so that it is only done there, and we don't have a chance of double counting?

Reply:
Stats are charged on each of the terminal paths of which only one will be called per stream - `onReset`, `onComplete`, and `onCancel`. `closeStream` on the other hand is not guaranteed to be called.